File: src/main/java/expensereport/ExpenseReport.java
package expensereport;

import java.util.ArrayList;
import java.util.List;

import static expensereport.Expense.Type.BREAKFAST;
import static expensereport.Expense.Type.DINNER;


public class ExpenseReport {
    private List<Expense> expenses = new ArrayList<Expense>();
    private int total;
    private int mealExpenses;

    public void printReport(ReportPrinter printer) {
        total = 0;
        mealExpenses = 0;

        printHeader(printer);

        totalUpExpenses();

        for (Expense expense : expenses) {
            String name = "TILT";
            switch (expense.type) {
                case DINNER:
                    name = "Dinner";
                    break;
                case BREAKFAST:
                    name = "Breakfast";
                    break;
                case CAR_RENTAL:
                    name = "Car Rental";
                    break;
            }
            printer.print(String.format("%s\t%s\t$%.02f\n",
                    ((expense.type == DINNER && expense.amount > 5000)
                            || (expense.type == BREAKFAST && expense.amount > 1000)) ? "X" : " ",
                    name, expense.amount / 100.0));
        }
        printTotals(printer);
    }

    private void totalUpExpenses() {
        for (Expense expense : expenses) {
            if (expense.type == BREAKFAST || expense.type == DINNER)
                mealExpenses += expense.amount;

            total += expense.amount;
        }
    }

    private void printTotals(ReportPrinter printer) {
        printer.print(String.format("\nMeal expenses $%.02f", mealExpenses / 100.0));
        printer.print(String.format("\nTotal $%.02f", total / 100.0));
    }

    private void printHeader(ReportPrinter printer) {
        printer.print("Expenses " + getDate() + "\n");
    }

    public void addExpense(Expense expense) {
        expenses.add(expense);
    }

    private String getDate() {
        return "9/12/2002";
    }
}

File: SwiftExpense/SwiftExpense/Example/ExpenseReport.swift
//
//  ExpenseReport.swift
//  SwiftExpense
//
//  Created by tskim on 2019/11/06.
//  Copyright © 2019 assin. All rights reserved.
//

import Foundation

class ExpenseReport {
    private var expenses = [Expense]()

    func printReport(printer: ReportPrinter) {
        var total = 0
        var mealExpenses = 0

        printHeader(printer: printer)
        for expense in expenses {

            if expense.type == .breakfast || expense.type == .dinner {
                mealExpenses += expense.amount
            }

            var name = "TILT"

            switch expense.type {
            case .dinner:
                name = "Dinner"
            case .breakfast:
                name = "Breakfast"
            case .carRental:
                name = "Car Rental"
            }
            
            printer.reportPrint(
                text: String(format: "%@\t%@\t$%.02f\n",
                    ((expense.type == .dinner && expense.amount > 5000) ||
                            (expense.type == .breakfast && expense.amount > 1000)) ? "X" : " ", name, Double(expense.amount) / 100.0))
            
            total += expense.amount
        }
        
        printer.reportPrint(text: String(format: "\nMeal expenses $%.02f", Double(mealExpenses) / 100.0))
        printer.reportPrint(text: String(format: "\nTotal $%.02f", Double(total) / 100.0))
    }
    
    func addExpense(expense: Expense) {
        expenses.append(expense)
    }
        
    private func printHeader(printer: ReportPrinter) {
        printer.reportPrint(text: "Expenses \(getDate())\n")
    }
    
    private func getDate() -> String {
        return "9/12/2002"
    }
}


